### PR TITLE
create github actions for deployment

### LINKED ISSUE #13: Create GitHub Actions Workflow for Deployment
We'd like to have the production site get updated automatically on every `push`/`merge` to the `main` branch.  It would be ideal if we could do this automatically using GitHub Actions.  [Docusaurus has docs for this](https://docusaurus.io/docs/deployment#triggering-deployment-with-github-actions) that someone should research and implement.  You can test it out in your own fork and make sure it works before we merge it here.

### PR DESCRIPTION
**Pull request by trangntt-016**
Fix #13 

Create .github/workflows/deployment.yaml to have the production site get updated automatically on every push/merge to the main branch.

Current progress:

* [x] deploy on GH Pages
* [x] update `CONTRIBUTING.md`

### CODE REVIEW

**Comment by ycechungAI** [comment]
I am going to test this by cloning your repo

File: .github/workflows/deployment.yaml
Comment (DukeManh) [inline .github/workflows/deployment.yaml:38]: I think we'll need a repo owner to add a public key to Github deploy keys and a private key to Github secrets so Docusaurus can run the deployment for us.
CC @humphd

File: .github/workflows/deployment.yaml
Comment (DukeManh) [inline .github/workflows/deployment.yaml:52]: Shouldn't this be `npm run deploy` so the it's get deployed and the production site gets updated?

File: .github/workflows/deployment.yaml
Comment (DukeManh) [inline .github/workflows/deployment.yaml:44]: This should be using `Seneca-ICTOER` email and username too right?

File: .idea/.gitignore
Comment (DukeManh) [inline .idea/.gitignore:1]: I don't think it's a good idea to commit 'Intellij'-specific files here because most of us are running VSCode.

File: .idea/IPC144.iml
Comment (DukeManh) [inline .idea/IPC144.iml:1]: These files too.

**Review by DukeManh — CHANGES_REQUESTED** [review]
OSD peer-review 🤝.

I'm new to Github actions too, left some comments, let me know what u think.

**Comment by ycechungAI** [comment]
She has demo it to me on November 9th around 1pm and the following is it running on a test repo with code

**Comment by ycechungAI** [comment]
issue-13_working_ycechung_Tracy

File: .idea/IPC144.iml
Comment (ycechungAI) [reply .idea/IPC144.iml:1]: i agree

File: .github/workflows/deployment.yaml
Comment (ycechungAI) [reply .github/workflows/deployment.yaml:44]: I think she was trying to deploy on vercel but she should make these changes

**Comment by humphd** [comment]
Nice! What's the URL for the live version it pushed to @trangntt-016's repo?  When this completes the end product should be a built version that's live GH Pages.  I think there's still more steps to do.

File: .github/workflows/deployment.yaml
Comment (ycechungAI) [reply .github/workflows/deployment.yaml:38]: I agree, but it seems you need to add it to docker setup if cause I'm trying it on act-cli running it on local

Comment (ycechungAI) [reply .github/workflows/deployment.yaml:38]: also local .env file like in Telescope like a for the search feature as well for my PR

**Comment by ycechungAI** [comment]
Based on Duke's comments yes a few tweaks here and there.

**Comment by DukeManh** [comment]
> What's the URL for the live version it pushed to @trangntt-016's repo?

It hasn't been deployed yet.

https://github.com/Seneca-ICTOER/IPC144/pull/48#discussion_r746305102

**Comment by humphd** [comment]
We should figure out how to connect this with #62 as well.  That can happen in a follow-up, once both pieces land.

@ycechungAI can you file an issue discussing what you think needs to happen in order to get the indexing outputs to get integrated into what we publish?

**Comment by humphd** [comment]
Similar thing for the work in #41. Later we'll want to include the PDF build step in our deployment pipeline, so we can host the downloadable PDF after each update.  cc @TueNguyen2911.  We should get an issue filed to connect this work and that.

File: .github/workflows/deployment.yaml
Comment (humphd) [inline .github/workflows/deployment.yaml:5]: Let's get rid of the pull request bits here and below, and only do this on pushes to `main`.  We do pull request previews with Vercel already, so we know if it's working.

File: .github/workflows/deployment.yaml
Comment (humphd) [inline .github/workflows/deployment.yaml:10]: You can get rid of this

File: .github/workflows/deployment.yaml
Comment (humphd) [inline .github/workflows/deployment.yaml:29]: We can drop this `if`

File: .github/workflows/deployment.yaml
Comment (humphd) [inline .github/workflows/deployment.yaml:32]: use @v2

File: .github/workflows/deployment.yaml
Comment (humphd) [inline .github/workflows/deployment.yaml:33]: Use @v2 here as well.

File: .github/workflows/deployment.yaml
Comment (humphd) [inline .github/workflows/deployment.yaml:35]: Update this to 16, which is the current LTS version (12 is old now).

File: .idea/IPC144.iml
Comment (trangntt-016) [reply .idea/IPC144.iml]: thanks, I've just removed `.idea` folders

File: .github/workflows/deployment.yaml
Comment (humphd) [reply .github/workflows/deployment.yaml:35]: Let's also add a `cache: 'yarn'` below this line, see https://github.com/actions/setup-node#caching-packages-dependencies

File: .github/workflows/deployment.yaml
Comment (humphd) [reply .github/workflows/deployment.yaml:38]: Once we're ready to merge this, we can do that

File: .github/workflows/deployment.yaml
Comment (humphd) [inline .github/workflows/deployment.yaml:52]: Newline after this (always end a file on a new line).

File: .github/workflows/deployment.yaml
Comment (trangntt-016) [reply .github/workflows/deployment.yaml]: > Shouldn't this be `npm run deploy` so the it's get deployed and the production site gets updated?

thanks, I've just changed it to `yarn deploy`  which performs similar functionality

**Review by humphd — CHANGES_REQUESTED** [review]
Getting close, this is great.

File: .github/workflows/deployment.yaml
Comment (trangntt-016) [reply .github/workflows/deployment.yaml:44]: > This should be using `Seneca-ICTOER` email and username too right?

thanks, I would like to keep it
`git config --global user.name ${{ github.actor }}` because it provides more flexibility than setting a fixed variable `Seneca-ICTOER`

Ref:
`github.actor` is the login of the user that initiated the workflow run (https://docs.github.com/en/actions/learn-github-actions/contexts)

File: .idea/.gitignore
Comment (trangntt-016) [reply .idea/.gitignore]: > I don't think it's a good idea to commit 'Intellij'-specific files here because most of us are running VSCode.

yeah right, I used Webstorm and it automatically generated the `.idea` file. Should have deleted it when committing.

File: .github/workflows/deployment.yaml
Comment (trangntt-016) [reply .github/workflows/deployment.yaml:38]: > I think we'll need a repo owner to add a public key to Github deploy keys and a private key to Github secrets so Docusaurus can run the deployment for us. CC @humphd



> Nice! What's the URL for the live version it pushed to @trangntt-016's repo? When this completes the end product should be a built version that's live GH Pages. I think there's still more steps to do.

Hi @humphd ,

thanks for your feedback. Agree with @DukeManh , and I also tested on my `test` branch and it worked after I added an SSH key in my repo's setting. I wonder if you could please generate a secret key `GH_PAGES_DEPLOY` and add it to `ICTOER` setting?

File: .idea/.gitignore
Comment (humphd) [reply .idea/.gitignore]: @trangntt-016 do you want to do another PR to add `.idea` to our https://github.com/Seneca-ICTOER/IPC144/blob/main/.gitignore?  Then other students that use it won't have this problem.

File: .github/workflows/deployment.yaml
Comment (humphd) [reply .github/workflows/deployment.yaml:38]: Can I ask that you document the steps that need to be done in `CONTRIBUTING.md`, and we can do that.  I want to make sure it's clear for future maintainers what needs to happen vs. just doing it once.

**Comment by ycechungAI** [comment]
> 
> 
> We should figure out how to connect this with #62 as well. That can happen in a follow-up, once both pieces land.
> 
> @ycechungAI can you file an issue discussing what you think needs to happen in order to get the indexing outputs to get integrated into what we publish?

Its part of #70 and ill take time to expand on it as it is not so clear on the website

**Comment by trangntt-016** [comment]
Hi @humphd ,

I've done a few changes based on your feedback. Here's the output
image

Please let me know if I need to have additional changes and lemme know when can I start with  #14  #62 #41. I'm so excited connecting with them

File: docusaurus.config.js
Comment (humphd) [inline docusaurus.config.js:8]: This is wrong for our case, please revert.

**Review by humphd — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (trangntt-016) [reply CONTRIBUTING.md:65]: > i.e., do we need to do this in our case, or are you documenting this for people setting it up on their forks?

yes, I'm documenting so that people can deploy locally on their forks

File: .gitignore
Comment (trangntt-016) [reply .gitignore:28]: > In the future, this type of change should be done in a separate PR, since it's unrelated to what you're fixing in this one. It's better to have more, smaller PRs, vs adding complexity and scope to existing ones.

noted thanks

File: CONTRIBUTING.md
Comment (humphd) [reply CONTRIBUTING.md:65]: OK.  Because that's not our primary use case here, I'd suggest you add a note that indicates that it's only necessary if you're testing on a fork.  Otherwise, focus on what this site/repo needs to do.

**Review by ycechungAI — COMMENTED** [review]
reviewed changes, but not sure about my algolia-scraper

**Comment by trangntt-016** [comment]
Hi @humphd ,

Please have a look if my PR is alright

Should I create a new PR to connect with #41 #62 , etc.?

File: .github/workflows/algolia-scraper.yml
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:1]: This should not be in here, since it's happening in #62.  Please remove all the algolia/search stuff from this PR and only do your deployment work.

File: config.json
Comment (humphd) [inline config.json:1]: This can be removed.

File: scrape.sh
Comment (humphd) [inline scrape.sh:1]: And this as well.

**Review by humphd — CHANGES_REQUESTED** [review]
Please make the changes, then squash and rebase your work so we can merge it today.

@CameronGray1210 we'll have to do some setup work and add secrets to the repo in order to make this possible.

File: .github/workflows/algolia-scraper.yml
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:1]: @trangntt-016 ^

File: scrape.sh
Comment (trangntt-016) [reply scrape.sh]: thanks, I've deleted it

File: config.json
Comment (trangntt-016) [reply config.json]: yup, I've deleted it

File: .github/workflows/algolia-scraper.yml
Comment (trangntt-016) [reply .github/workflows/algolia-scraper.yml]: deleted it

**Review by humphd — APPROVED** [review]

**Comment by humphd** [comment]
I'm going to merge this now, and it will fail until we install the secrets in the repo, however, we can ignore it.  FYI @CameronGray1210

**Comment by humphd** [comment]
Current merge failed with:

```
Error: The ssh-private-key argument is empty. Maybe the secret has not been configured, or you are using a wrong secret name in your workflow file.
```

We can ignore this for now.

**Review by nguyenhung15913 — COMMENTED** [review]
Do think if this should available to older node version, since I think a lot of people have not updated it to version 16.0.0 yet?

File: .github/workflows/deployment.yaml
Comment (trangntt-016) [reply .github/workflows/deployment.yaml:14]: yeah that's a good idea, I didn't think about this. Maybe you can create an issue and then I'll make another PR to fix this as this one is already closed.

Comment (humphd) [reply .github/workflows/deployment.yaml:14]: This deployment step only happens on GitHub Actions, where we control the node version.  This doesn't relate to what a user has on their machine.

**Pull request merged**
